src/builtin_complete.cpp

```diff
@@ -336,7 +336,11 @@ int builtin_complete(parser_t &parser, io_streams_t &streams, wchar_t **argv) {
         parser.libdata().transient_commandlines.push_back(do_complete_param);
         cleanup_t remove_transient([&] { parser.libdata().transient_commandlines.pop_back(); });
 
-        if (parser.libdata().builtin_complete_recursion_level < 1) {
+        // Allow at most two recursive calls to complete (#3474).
+        if (parser.libdata().builtin_complete_recursion_level >= 2) {
```

Comment:
Is there a particular reason to limit this to 2? I'd personally give it the same limit as wrapping (24, defined directly in complete.cpp::walk_wrap_chain_recursive).

(and even that is mostly there as a defence against infinite loops, and isn't the full story - see #5638)

Reply:
Not really any reason for a limit of 2 I think.
I have changed it to a hardcoded 24 now.

Reply:
The wrap chain handling should probably also print an error message when the limit is reached, to tell the user that there is probably a cycle.

Reply:
Yeah, that's probably correct. Tho we don't want to spam the error.